crates/astria-conductor/src/executor/mod.rs

```diff
@@ -447,29 +460,49 @@ impl Executor {
             )
         };
 
-        let update_type =
-            if let Some(block) = self.blocks_pending_finalization.remove(&block_number) {
-                info!(
-                    block_number,
-                    "found pending block; updating state but not not re-executing it"
-                );
-                Update::OnlyFirm(block)
-            } else {
-                info!(
-                    block_number,
-                    "did not find pending block; executing firm block"
-                );
-                // The parent hash of the next block is the hash of the block at the current head.
-                let parent_hash = self.state.firm_hash();
-                let executed_block = self
-                    .execute_block(client.clone(), parent_hash, executable_block)
-                    .await
-                    .wrap_err("failed to execute block")?;
-                self.does_block_response_fulfill_contract(ExecutionKind::Firm, &executed_block)
-                    .wrap_err("execution API server violated contract")?;
-                Update::ToSame(executed_block)
-            };
-        self.update_commitment_state(client.clone(), update_type)
+        let update = if self.should_execute_firm_block() {
+            let parent_hash = self.state.firm_hash();
+            let executed_block = self
+                .execute_block(client.clone(), parent_hash, executable_block)
+                .await
+                .wrap_err("failed to execute block")?;
+            self.does_block_response_fulfill_contract(ExecutionKind::Firm, &executed_block)
+                .wrap_err("execution API server violated contract")?;
+            Update::ToSame(executed_block)
+        } else if let Some(block) = self.blocks_pending_finalization.remove(&block_number) {
+            info!(
+                block_number,
+                "found pending block; updating state but not not re-executing it"
+            );
+            Update::OnlyFirm(block)
+        } else {
+            // XXX: This case should never be reached because the firm block *must* exist in the
+            // cache - either due to being pre-populated at startup (via
+            // `populate_blocks_pending_finalization`), or during normal operation (as part of
+            // `execute_soft`).
+            // This code is here as a fall-back mechanism in case there is a bug.
+            error!(
+                block_number,
+                "pending block not found for block number in cache. THIS SHOULD NOT HAPPEN. \
+                 Trying to fetch the already-executed block from the rollup before giving up."
+            );
+            match client.clone().get_block(block_number).await {
```

Comment:
@joroshiba Flagging this new logic here: we could forego the entire logic introduced in https://github.com/astriaorg/astria/pull/980 and just try and fetch the missing blocks (those pending finalization) when we actually need them.

Would make startup cheaper at the cost of extra network requests.